src/storage/src/source/generator/counter.rs

```diff
@@ -22,14 +27,50 @@ impl Generator for Counter {
         _seed: Option<u64>,
     ) -> Box<dyn Iterator<Item = (usize, GeneratorMessageType, Row, i64)>> {
         let mut counter = 0;
-        Box::new(iter::repeat_with(move || {
-            counter += 1;
-            (
-                0,
-                GeneratorMessageType::Finalized,
-                Row::pack_slice(&[Datum::Int64(counter)]),
-                1,
-            )
-        }))
+        let max_cardinality = self.max_cardinality;
+        Box::new(
+            iter::repeat_with(move || {
+                let to_retract = match max_cardinality {
+                    Some(max) => {
+                        if max <= counter {
+                            Some(counter - max + 1)
+                        } else {
+                            None
+                        }
+                    }
+                    None => None,
+                };
+                counter += 1;
```

Comment:
If you moved this above to_retract you could remove the +1 in the to_retract calculation.

Reply:
Yes, but then the if statement would have to read `if max <= counter - 1`, which doesn't seem like a net win.